design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+        annotations:
+          crossplane.io/external-name-suffix: "-highmem"
+      spec:
+        forProvider:
+          clusterSelector:
+            matchConfiguration: true
+          autoscaling:
+            enabled: true
+            minNodeCount: 0
+            maxNodecount: 100
+          config:
+            diskType: PD_SSD
+            machineType: n1-highmem-32
+          version: 1.16
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    patch:
+    - fromClaimField: ".spec.version"
+      toResourceField: ".spec.forProvider.version"
+```
+
+The below `ResourceConfiguration` is produced by the above `MySQLInstance` and
+`ResourceConfigurationClass`:
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceConfiguration
+metadata:
+  name: default-cluster-fdjk2
+  annotations:
+    kubernetescluster.compute.crossplane.io/v1alpha1: bindable
+spec:
+  resourceSelector:
+    matchLabels:
+      resourceclaim.crossplane.io/namespace: default
+      resourceclaim.crossplane.io/name: cluster
+      resourceclaim.crossplane.io/uid: 126c5310-35b1-11ea-978f-2e728ce88125
+  claimRef:
+    apiVersion: compute.example.org/v1alpha1
+    kind: KubernetesCluster
+    namespace: default
+    name: cluster
+  classRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: ResourceConfigurationClass
+    name: gke-cluster
+  writeConnectionSecretsToRef:
+    namespace: crossplane-system
+    name: 126c5310-35b1-11ea-978f-2e728ce88125
+status:
+  resources:
+  - apiVersion: container.gcp.crossplane.io/v1beta1
+    kind: GKECluster
+    name: default-cluster-d82nd
+    bindingPhase: Bound
+  - apiVersion: container.gcp.crossplane.io/v1alpha3
+    kind: NodePool
+    name: default-cluster-9dm3v
+    bindingPhase: Bound
+  - apiVersion: container.gcp.crossplane.io/v1alpha3
+    kind: NodePool
+    name: default-cluster-m23f3
+    bindingPhase: Bound
+  - apiVersion: container.gcp.crossplane.io/v1alpha3
+    kind: NodePool
+    name: default-cluster-2if9d
+    bindingPhase: Bound
+  bindingPhase: Bound
+```
+
+Note that the `ResourceConfiguration` binds to the `KubernetesCluster` claim,
+while the dynamically provisioned managed resources bind in turn to the
+`ResourceConfiguration`, thus creating a `resource claim -> resource
+configuration -> managed resource(s)` binding relationship. The below
+`GKECluster` binds to a `ResourceConfiguration` (per its `configurationRef` and
+`bindingPhase`) rather than to a resource claim:
+
+```yaml
+apiVersion: compute.gcp.crossplane.io/v1beta1
+kind: GKECluster
+metadata:
+  name: default-cluster-ffm8f
+  labels:
+    resourceclaim.crossplane.io/namespace: default
+    resourceclaim.crossplane.io/name: cluster
+    resourceclaim.crossplane.io/uid: 126c5310-35b1-11ea-978f-2e728ce88125
+spec:
+  forProvider:
+    # Omitted for brevity.
+  classRef:
+    apiVersion: infrastructure.crossplane.io/v1beta1
+    kind: ResourceConfigurationClass
+    name: gke-cluster
```

Comment:
Apologies if this was mentioned in this write-up and I missed it, but would `ResourceConfigurationClass` fully replace strongly typed classes, or do we intend to keep them around? I presume we could keep them around for some time and then retire them if all users were creating a single resource `ResourceConfigurationClass` instead of using strongly typed classes.

Reply:
It's only mentioned in a note under the backward compatibility heading at the moment. It needs a little more thought, but I suspect we would be able to keep them around. I do imagine we'd want to deprecate them and remove them after a few releases though, if only to keep the mental model as simple as possible.